### PR TITLE
Upstream some binary-compatibility logic

### PR DESCRIPTION
**Pull request by tbkka**
A couple of  compatibility checks to preserve behavior for existing binaries on Apple platforms.

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please test

**Review by mikeash — APPROVED** [review]

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
@swift-ci Please test macOS Platform

**Comment by tbkka** [comment]
@swift-ci Please test macOS platform

**Comment by tbkka** [comment]
@swift-ci Please test macOS Platform

**Comment by tbkka** [comment]
@ktoso -- It looks like I needed a small test adjustment in order to upstream the bincompat check for executor assertion behavior.  Any concerns?

**Comment by tbkka** [comment]
@swift-ci Please test

File: test/Distributed/Runtime/distributed_actor_custom_executor_availability.swift [hunk 1]
```diff
@@ -38,7 +38,7 @@ import Distributed
       }
 
       tests.test("5.7 actor, 5.9 executor property => no custom executor") {
-        expectCrashLater(withMessage: "Incorrect actor executor assumption; Expected MainActor executor")
+        expectCrashLater()
```
Comment (tbkka) [inline test/Distributed/Runtime/distributed_actor_custom_executor_availability.swift:41]: The new semantics generate the crash in the `swift_task_isCurrentExecutorImpl` function, which occurs before the message is emitted.

**Comment by Kyle-Ye** [comment]
Some Notes on this PR: @tbkka 

1. This represents a behavior-breaking change when building with the Xcode/Swift toolchain, which could cause some applications to crash. While I support the new behavior, it would be beneficial to have this information documented in the vendor's official release notes (e.g., Xcode 16 release notes).

2. Regarding the line: `const dyld_build_version_t fall_2024_os_versions = {0xffffffff, 0x007e80000};`

I recall that `dyld_build_version_t` is defined as a struct containing a uint32 platform and a uint32 version. Therefore, I believe the latter value should be `0x07e80000` instead of `0x007e80000`, as the current value exceeds 32 bits. Though most compilers will likely truncate this value correctly, it would be more accurate to use the proper 32-bit representation.

**Pull request merged**